src/persist-client/src/internal/compact.rs

```diff
@@ -832,6 +825,104 @@ where
+    cfg: PersistConfig,
+    metrics: Arc<Metrics>,
+    _phantom: PhantomData<T>,
+}
+
+impl<T> Clone for CompactionFilter<T> {
+    fn clone(&self) -> Self {
+        Self {
+            cfg: self.cfg.clone(),
+            metrics: Arc::clone(&self.metrics),
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<T: PartialOrder> CompactionFilter<T> {
+    fn new(cfg: PersistConfig, metrics: Arc<Metrics>) -> Self {
+        Self {
+            cfg,
+            metrics,
+            _phantom: Default::default(),
+        }
+    }
+
+    fn should_compact(&self, req: &CompactReq<T>) -> bool {
+        let blob_target_size = self.cfg.dynamic.blob_target_size();
+
+        if !Self::may_have_logical_compaction(req)
+            && !Self::may_have_physical_compaction(req, blob_target_size)
+        {
+            self.metrics.compaction.filter.drop_unneeded.inc();
+            return false;
+        }
+
+        // At this point, we know our req contains batches that may benefit from
+        // logical and/or physical compaction. Apply some heuristics to see whether
+        // there's enough work in some dimension (# batches, # parts, # updates)
+        // to be worthwhile.
+
+        if req.inputs.len() >= self.cfg.dynamic.compaction_heuristic_min_inputs() {
+            self.metrics.compaction.filter.pass_batches.inc();
+            return true;
+        }
+
+        if req.inputs.iter().map(|x| x.parts.len()).sum::<usize>()
+            >= self.cfg.dynamic.compaction_heuristic_min_parts()
+        {
+            self.metrics.compaction.filter.pass_parts.inc();
+            return true;
+        }
+
+        if req.inputs.iter().map(|x| x.len).sum::<usize>()
+            >= self.cfg.dynamic.compaction_heuristic_min_updates()
+        {
+            self.metrics.compaction.filter.pass_updates.inc();
+            return true;
+        }
+
+        self.metrics.compaction.filter.drop_no_match.inc();
+        false
+    }
+
+    fn may_have_logical_compaction(req: &CompactReq<T>) -> bool {
+        // if the req's `since` is beyond any input's, there may be logical compaction to do
+        req.inputs
+            .iter()
+            .any(|x| PartialOrder::less_than(x.desc.since(), req.desc.since()))
+    }
+
+    fn may_have_physical_compaction(req: &CompactReq<T>, blob_target_size: usize) -> bool {
+        // physical compaction is beneficial when there's >1 part below the
+        // target size, so that the space used by the small ones may be merged
+        // together. having zero or one parts below the target size means that
+        // overall the batches are already as tightly packed as they can be
+        // given the current level of logical compaction.
+        let parts_below_target_size = req
+            .inputs
+            .iter()
+            .flat_map(|x| x.parts.iter())
+            .filter(|x| x.encoded_size_bytes < blob_target_size)
+            .count();
+        if parts_below_target_size > 1 {
```

Comment:
hmm still pondering this. I think this is correct for whether physical compaction would change anything, but I don't think it helps us avoid huge batches in the long `since` hold scenario. In that case, I'd expect us to get a compaction req for >= 2 batches that each have a trailing part < `blob_target_size` (since it's unlikely a batch would be exactly a multiple of `blob_target_size`)